clash-prelude/src/Clash/Prelude/Bundle.hs

```diff
@@ -118,53 +122,62 @@ class Bundle a where
   -- @
   -- __unbundle__ :: 'Signal' dom 'Clash.Sized.BitVector.Bit' -> 'Signal' dom 'Clash.Sized.BitVector.Bit'
   -- @
-  unbundle :: Signal dom a -> Unbundled dom a
+  unbundle :: f a -> res
 
   {-# INLINE unbundle #-}
-  default unbundle :: (Unbundled dom a ~ Signal dom a)
-                   => Signal dom a -> Unbundled dom a
-  unbundle s = s
-
-instance Bundle ()
-instance Bundle Bool
-instance Bundle Integer
-instance Bundle Int
-instance Bundle Float
-instance Bundle Double
-instance Bundle (Maybe a)
-instance Bundle (Either a b)
-
-instance Bundle Bit
-instance Bundle (BitVector n)
-instance Bundle (Index n)
-instance Bundle (Fixed rep int frac)
-instance Bundle (Signed n)
-instance Bundle (Unsigned n)
+  default unbundle :: (f a ~ res) => f a -> res
+  unbundle = id
+
+instance Bundle f () (f ())
+instance Bundle f Bool (f Bool)
+instance Bundle f Integer (f Integer)
+instance Bundle f Int (f Int)
+instance Bundle f Float (f Float)
+instance Bundle f Double (f Double)
+instance Bundle f (Maybe a) (f (Maybe a))
+instance Bundle f (Either a b) (f (Either a b))
+
+instance Bundle f Bit (f Bit)
+instance Bundle f (BitVector n) (f (BitVector n))
+instance Bundle f (Index n) (f (Index n))
+instance Bundle f (Fixed rep int frac) (f (Fixed rep int frac))
+instance Bundle f (Signed n) (f (Signed n))
+instance Bundle f (Unsigned n) (f (Unsigned n))
 
 -- | __N.B.__: The documentation only shows instances up to /3/-tuples. By
 -- default, instances up to and including /12/-tuples will exist. If the flag
 -- @large-tuples@ is set instances up to the GHC imposed limit will exist. The
 -- GHC imposed limit is either 62 or 64 depending on the GHC version.
 deriveBundleTuples ''Bundle ''Unbundled 'bundle 'unbundle
 
-instance KnownNat n => Bundle (Vec n a) where
-  type Unbundled t (Vec n a) = Vec n (Signal t a)
-  -- The 'Traversable' instance of 'Vec' is not synthesizable, so we must
-  -- define 'bundle' as a primitive.
+instance {-# OVERLAPPABLE #-} (KnownNat n, Applicative f, Traversable f) => Bundle f (Vec n a) (Vec n (f a)) where
+  type Unbundled f (Vec n a) = Vec n (f a)
+  bundle   = traverse# id
+  unbundle = sequenceA . fmap lazyV
+
+instance KnownNat n => Bundle (Signal dom) (Vec n a) (Vec n (Signal dom a)) where
+  type Unbundled (Signal dom) (Vec n a) = Vec n (Signal dom a)
```

Comment:
What's the point of the more specific `OVERLAPPING` instances?
Does it lead to better performance?

Reply:
Better run-time performance, no space leakage, and with the ability filter out the applications of the methods also better Haskell-to-HDL compile-time performance.

Reply:
Ah, sorry, I thought the question was about all of the `Bundle` instances. So in the `Vec` case, the overlapping instances are there so can filter out the applications of `vecBundle#` (resp `vecBundleD#`) in GHC2Core. We want this because `traverse# :: Applicative f => (a -> f b) -> Vec n a -> f (Vec n b)` is always unrolled by: https://github.com/clash-lang/clash-compiler/blob/5f5739a775b9713f971ba33557771659fc1eb2aa/clash-lib/src/Clash/Normalize/Transformations/Reduce.hs#L199-L205